homeassistant/components/tesla/switch.py

```diff
@@ -12,136 +12,134 @@
+        _LOGGER.debug("Enable charging: %s", self.name)
         await self.tesla_device.start_charge()
 
     async def async_turn_off(self, **kwargs):
         """Send the off command."""
-        _LOGGER.debug("Disable charging for: %s", self._name)
+        _LOGGER.debug("Disable charging for: %s", self.name)
         await self.tesla_device.stop_charge()
 
     @property
     def is_on(self):
         """Get whether the switch is in on state."""
-        return self._state == STATE_ON
+        if self.tesla_device.is_charging() is None:
+            return None
+        return self.tesla_device.is_charging() == STATE_ON
 
     async def async_update(self):
         """Update the state of the switch."""
-        _LOGGER.debug("Updating state for: %s", self._name)
+        _LOGGER.debug("Updating state for: %s", self.name)
         await super().async_update()
-        self._state = STATE_ON if self.tesla_device.is_charging() else STATE_OFF
 
 
 class RangeSwitch(TeslaDevice, SwitchEntity):
     """Representation of a Tesla max range charging switch."""
 
-    def __init__(self, tesla_device, controller, config_entry):
-        """Initialise the switch."""
-        self._state = None
-        super().__init__(tesla_device, controller, config_entry)
-
     async def async_turn_on(self, **kwargs):
         """Send the on command."""
-        _LOGGER.debug("Enable max range charging: %s", self._name)
+        _LOGGER.debug("Enable max range charging: %s", self.name)
         await self.tesla_device.set_max()
 
     async def async_turn_off(self, **kwargs):
         """Send the off command."""
-        _LOGGER.debug("Disable max range charging: %s", self._name)
+        _LOGGER.debug("Disable max range charging: %s", self.name)
         await self.tesla_device.set_standard()
 
     @property
     def is_on(self):
         """Get whether the switch is in on state."""
-        return self._state
+        if self.tesla_device.is_maxrange() is None:
+            return None
+        return bool(self.tesla_device.is_maxrange())
 
     async def async_update(self):
         """Update the state of the switch."""
-        _LOGGER.debug("Updating state for: %s", self._name)
+        _LOGGER.debug("Updating state for: %s", self.name)
         await super().async_update()
-        self._state = bool(self.tesla_device.is_maxrange())
 
 
 class UpdateSwitch(TeslaDevice, SwitchEntity):
     """Representation of a Tesla update switch."""
 
-    def __init__(self, tesla_device, controller, config_entry):
+    def __init__(self, tesla_device, coordinator, controller):
         """Initialise the switch."""
-        self._state = None
-        tesla_device.type = "update switch"
-        super().__init__(tesla_device, controller, config_entry)
-        self._name = self._name.replace("charger", "update")
-        self.tesla_id = self.tesla_id.replace("charger", "update")
+        super().__init__(tesla_device, coordinator)
+        self.controller = controller
```

Comment:
i would create your own data update coordinator class that stores the controller instance so you not passing it around separately. then you can do self.coordinator.controller to get your reference

https://github.com/home-assistant/core/blob/dev/homeassistant/components/roku/__init__.py#L111

Reply:
This is the only sensor that will actually need access to the controller. The controller was basically the DataUpdateCoordinator for the component. I'm not sure it's worth extending the DataUpdateCoordinator for a single use case but could be convinced.

Reply:
Sorry, just realized you started implementing it. If you're already adding the code, then I'm ok with just taking your commits.